ctk/common/src/test/kotlin/org/codice/compilance/verification/core/SamlDefinedIdentifiersVerifierSpec.kt

```diff
@@ -0,0 +1,281 @@
+                    |  </samlp:Status>
+                    |  <saml:Assertion ID="id" IssueInstant="$now" Version="$VERSION">
+                    |    <saml:Issuer
+                    |    $extraIdentifierAttribute
+                    |    Format="$identifierFormat">$identifierValue</saml:Issuer>
+                    |    <saml:AttributeStatement>
+                    |      <saml:Attribute Name="$attributeName" NameFormat="$attributeFormat" />
+                    |    </saml:AttributeStatement>
+                    |  </saml:Assertion>
+                    |</samlp:Response>
+                    """.trimMargin())
+        }
+
+        /* 8.2 URI/Basic name attribute formats */
+        "valid 'unspecified' attribute name" {
+            createResponse(
+                    attributeName = "Unspecified Name",
+                    attributeFormat = ATTRIBUTE_NAME_FORMAT_UNSPECIFIED).let {
+                SamlDefinedIdentifiersVerifier(it).verify()
+            }
+        }
+
+        // No invalid test for 'unspecified' format; other than what is allowed in xml
+
+        "valid URI attribute name" {
+            createResponse(
+                    attributeName = "validURI.com",
+                    attributeFormat = ATTRIBUTE_NAME_FORMAT_URI).let {
+                SamlDefinedIdentifiersVerifier(it).verify()
+            }
+        }
+
+        "invalid URI attribute name" {
+            createResponse(
+                    attributeName = "Whitespace Not Allowed In URI",
+                    attributeFormat = ATTRIBUTE_NAME_FORMAT_URI).let {
+                shouldThrow<SAMLComplianceException> {
+                    SamlDefinedIdentifiersVerifier(it).verify()
+                }.message?.shouldContain(SAMLCore_8_2_2_a.message)
+            }
+        }
+
+        "valid Basic attribute name" {
+            createResponse(
+                    attributeName = "BasicName",
+                    attributeFormat = ATTRIBUTE_NAME_FORMAT_BASIC).let {
+                SamlDefinedIdentifiersVerifier(it).verify()
+            }
+        }
+
+        "invalid Basic attribute name" {
+            createResponse(
+                    attributeName = "Whitespace Not Allowed In Attribute Name",
+                    attributeFormat = ATTRIBUTE_NAME_FORMAT_BASIC).let {
+                shouldThrow<SAMLComplianceException> {
+                    SamlDefinedIdentifiersVerifier(it).verify()
+                }.message?.shouldContain(SAMLCore_8_2_3_a.message)
+            }
+        }
+
+        /* 8.3 Name Identifier Format Identifiers */
+        "valid Email name identifier" {
+            createResponse(
+                    identifierValue = "example-email@domain.com",
+                    identifierFormat = NAME_ID_FORMAT_EMAIL).let {
+                SamlDefinedIdentifiersVerifier(it).verify()
+            }
+        }
+
+        "invalid Email name identifier (multiple '@'s)" {
```

Comment:
You should combine all 9 invalid email tests into one
```
"invalid Email name identifier" {
            forAll("example@email@domain.com", "example-email.domain.com" etc ....) { email ->
                createResponse(
                     identifierValue = email,
                     identifierFormat = NAME_ID_FORMAT_EMAIL).let {
                 shouldThrow<SAMLComplianceException> {
                     SamlDefinedIdentifiersVerifier(it).verify()
                 }.message?.shouldContain(SAMLCore_8_3_2_a.message)
             }
        }
```

You can also add all the email to a list and do `forAll(invalidEmailList) { email ->`

Reply:
This is an awesome solution, I'm just worried about the information I would be losing doing it this way. It won't tell you why you failed; just that you failed on a certain email address.

Reply:
That seems fair to me 👍 I mean this is a unit test class, so reducing the amount of tests in here doesn't add just a ton of value.